ASoC: SOF: Start converting ipc.c to be IPC version neutral

ujfalusi opened the pull request:
Hi,

As a first step in finalizing the core's generic code to be IPC version neutral, the first step is to convert the high level IPC implementations to be neutral via callbacks.

The ipc.c is not going to be yet cleaned as the remaining code is used by the loader.c.

This series takes care of:
tx_message (send an IPC message to firmware)
  reply handling (handling the reply from firmware to a send message)
rx_message (notification, or firmware initiated messages)
set_get_data ('large' message handling for sending and receiving)
ipc3 kcontrol data handling via set_get_data

I have two additional planned PRs:
- global fw_ready ops removal and replacement with IPC specific implementations
- firmware loader rework to be IPC neutral.

Regards,
Peter

ujfalusi commented:
Interesting, I'm running sparse check as part of my local builds and I have not seen this error, I'll check and fix it.

plbossart commented:
> Interesting, I'm running sparse check as part of my local builds and I have not seen this error, I'll check and fix it.

you may want to rebase, differences between local and github checks are typical on base branch differences.

ranj063 commented on sound/soc/sof/sof-priv.h:
```diff
@@ -378,11 +378,25 @@ struct ipc_pcm_ops;
+ * @rx_msg:	Function pointer for handling a received message
+ * @set_get_data: Function pointer for set/get data, supporting large messages
+ * @get_reply:	Function pointer for fetching the reply to
+ *		sdev->ipc->msg.replay_data
```
typo reply

RanderWang replied:
gamer ? I used to watch replay

ujfalusi replied:
:D @RanderWang, thanks for spotting it!

ranj063 commented on sound/soc/sof/ipc3.c:
```diff
@@ -325,8 +325,8 @@ static int sof_ipc3_ctx_ipc(struct snd_sof_dev *sdev, int cmd)
-	return sof_ipc_tx_message(sdev->ipc, &pm_ctx, sizeof(pm_ctx),
-				  &reply, sizeof(reply));
+	return sof_ipc3_tx_msg(sdev, &pm_ctx, sizeof(pm_ctx),
+			       &reply, sizeof(reply), false);
```
this will fit in one line?

ujfalusi replied:
I try to avoid lines longer than 90 characters, it would be 93, I rather keep it like this. Yes, I know that 100 is the checkpatch limit.

ranj063 commented on sound/soc/sof/ipc4-ops.h:
```diff
@@ -0,0 +1,16 @@
+/* SPDX-License-Identifier: (GPL-2.0-only OR BSD-3-Clause) */
```
typo in commit message "sot he"

ranj063 commented on sound/soc/sof/ipc4-ops.h:
```diff
@@ -0,0 +1,16 @@
+ * This file is provided under a dual BSD/GPLv2 license.  When using or
+ * redistributing this file, you may do so under either license.
+ *
+ * Copyright(c) 2021 Intel Corporation. All rights reserved.
```
2022

ranj063 reviewed (commented):
looks awesome @ujfalusi. just a few typos.

RanderWang commented on sound/soc/sof/ipc.c:
```diff
@@ -1061,8 +1015,8 @@ struct snd_sof_ipc *snd_sof_ipc_init(struct snd_sof_dev *sdev)
-	    !ipc->ops->tplg->control) {
+	if (!ipc->ops->get_reply || !ipc->ops->pcm || !ipc->ops->tplg ||
+	    !ipc->ops->tplg->widget || !ipc->ops->tplg->control) {
 		dev_err(sdev->dev, "Invalid IPC ops\n");
```
a little crowded here ? more and more if

ujfalusi replied:
@RanderWang, yes it is. I will have this patch coming in later PR: https://github.com/ujfalusi/sof-linux/commit/d058116feaa04de914a1358bdd3c05c5f24b35cd

RanderWang reviewed (commented):
it looks very good

ujfalusi commented:
> it looks very good

Thank you!

ujfalusi commented:
oh, forgot the change log

Changes since v1:
- fix sparse warning
- corect typos

ranj063 reviewed (dismissed):
Thank you @ujfalusi. Looks great!

ujfalusi commented on sound/soc/sof/ipc4.c:
```diff
@@ -96,6 +96,32 @@ void sof_ipc4_process_reply(struct snd_sof_dev *sdev, u32 msg)
+	ipc4_reply = msg->reply_data;
+
+	if (ipc4_reply->primary & SOF_IPC4_REPLY_STATUS_MASK)
+		return ipc4_reply->primary & SOF_IPC4_REPLY_STATUS_MASK;
```
I think we should return `-EINVAL` for now and add an error code decoder later.

ujfalusi commented:
Changes since v2:
- return -EINVAL for now in case of an error status received in reply for IPC4.

RanderWang reviewed (dismissed):
LGTM

ujfalusi commented:
Changes since v3 (sorry):
- one new patch on top to convert the IPC4 status codes to standard errno numbers and remove the local ipc4.h while at it.

@ranj063, @RanderWang, unfortunately this push dropped your approval :(

ranj063 reviewed (dismissed):
@ujfalusi looks good but I wonder how we're going to handle upstreaming the last patch. I feel like it has to be a fix up because we're removing ipc4.h

plbossart commented on sound/soc/sof/sof-priv.h:
```diff
@@ -378,11 +378,25 @@ struct ipc_pcm_ops;
+
+	int (*tx_msg)(struct snd_sof_dev *sdev, void *msg_data, size_t msg_bytes,
+		      void *reply_data, size_t reply_bytes, bool no_pm);
+	void (*rx_msg)(struct snd_sof_dev *sdev);
```
any reason why the rx_msg cannot report an error with the void return type?

ujfalusi replied:
The existing `snd_sof_ipc_msgs_rx()` is void and I'm following that.
When we receive a message it does not really have status, it is a notification and we just place it to some place where higher level code can handle it.

plbossart commented on sound/soc/sof/sof-priv.h:
```diff
@@ -378,11 +378,25 @@ struct ipc_pcm_ops;
+ *
+ * @tx_msg:	Function pointer for sending an IPC message
+ * @rx_msg:	Function pointer for handling a received message
+ * @set_get_data: Function pointer for set/get data, supporting large messages
```
I don't fully get the difference or dependencies with tx_msg, and what 'large' means.

ujfalusi replied:
tx_msg can only transmit a message which can fit under the max ipc message size (the mailbox size technically)
The set_get_data can break down _certain_ type of messages into smaller tx_msg chunks and the handling of this is really message and IPC specific (for ipc3 we do an in-place tx/rx for example)

Initially I wanted to have generic it does all tx_msg, but it started to get really ugly when trying to handle certain types of messages in a different way and to figure out the buffers, pointers, tx part, reply part and handle multiple tx within a single tx. It is just a mess.

plbossart replied:
I meant: does 'set_get_data' internally call 'tx_msg', if yes that'd be worthy of a comment.

ujfalusi replied:
OK, I'll mention this.

ujfalusi replied:
@plbossart, at the moment both ipc3 and ipc4 internally uses their `*_tx_msg_unlocked` to break down a large message, but it is not a mandatory thing to do. In theory if we have some other IPC which can just send a 'large' message as it is, then it is not going to use the tx_msg.
So it is kind of IPC dependent how the large data is handled.

plbossart commented on sound/soc/sof/ipc3.c:
```diff
@@ -9,6 +9,312 @@
 
 #include "sof-priv.h"
 #include "ipc3-ops.h"
+#include "ops.h"
```
the commit message is a bit confusing. This could be understood as duplicating the code in ipc.c, but I think what you mean is that the code in ipc.c will be removed at a later point? make-then-break, yes?

ujfalusi replied:
Basically I'm using the existing code in ipc.c (which is the ipc3 implementation) to implement the new ops. The ops do require a bit of modification from the 'open-coded' versions due to the obsification, but the logic is intact.

Later when we have the ops in place, we switch to them and the old code can be removed.

plbossart commented on sound/soc/sof/ipc3.c:
```diff
@@ -316,6 +318,105 @@ static int sof_ipc3_tx_msg(struct snd_sof_dev *sdev, void *msg_data, size_t msg_
+
+	return ret;
+}
+
```
so the code also comes from ipc.c?

ujfalusi replied:
Well, it is based on the ipc.c implementation, but changed to fit to the new generic API split.

plbossart reviewed (changes requested):
The changes look fine, no real objections on the content, but I would have taken a different path with the patches to minimize the addition/removal of code that's going to be massively problematic for backports/fixes.

Can we instead
a) split ipc.c in two parts, ipc.c would use the ipc3 specific stuff and ipc-common what needs to remain
b) introduce new headers/helpers.
c) use the helpers for ipc.c

and in a separate PR add IPC4 stuff. This will make it easier to review IMHO.

ujfalusi commented:
> @ujfalusi looks good but I wonder how we're going to handle upstreaming the last patch. I feel like it has to be a fix up because we're removing ipc4.h

For upstream we anyway need to redo the IPC4 patches, there is no way around it. We have so much churn on top that we need to get it right for the first series.

ujfalusi commented:
> The changes look fine, no real objections on the content, but I would have taken a different path with the patches to minimize the addition/removal of code that's going to be massively problematic for backports/fixes.

This is true, but I don't see how can it be avoided. We will need IPC3 implementations (the one we have currently), IPC4 implementation and a common, generic code to make all this to work.
We do need to have changes in the IPC3 code in any case since we slice it up to have a sensible (I hope) layer where IPC implementation details can be hidden.

> Can we instead
> a) split ipc.c in two parts, ipc.c would use the ipc3 specific stuff and ipc-common what needs to remain
> b) introduce new headers/helpers.
> c) use the helpers for ipc.c
> 
> and in a separate PR add IPC4 stuff. This will make it easier to review IMHO.

My issue with this is that we already **have** ipc4.c and if we do any change in the IPC3 path it is logical to fix up the IPC4 at the same time in a topological way. I have separated the IPC4 patches on purpose, I don't have mixed ones just to avoid a mess for upstreaming.

The steps I do is:
a) introduce the ops and their purpose
b) add ipc3 implementations of them (using the current ones as template)
c) add ipc4 implementations of them (they are mostly new code as we don't really have anything for ipc4)
d) switch to the ops and drop the old code (removing the duplicated ipc3 code which is now obsolete)

It is possible to:
a) introduce the ops and their purpose
b) add ipc3 implementations of them (using the current ones as template)
c) switch to the ops and drop the old code (removing the duplicated ipc3 code which is now obsolete)
d) add ipc4 implementations of them (they are mostly new code as we don't really have anything for ipc4)

But at step 'c' it all looks broken (still compiles) since the ipc4.c is out of sync.

For upstreaming we will need to redo (squash, fixup, etc) the ipc4 for sure. There we can cleanly introduce the ops, ipc3 and all the preparatory things, then add the ipc4 implementations of the ops, and that's basically  all.

plbossart commented:
@ujfalusi We haven't sent the patches upstream, so what we can do is a PR with a big revert of all IPC4 code in topic/sof-dev, and restart with a clean slate.

XiaoyunWu6666 commented:
SOFCI TEST

ujfalusi commented:
> @ujfalusi We haven't sent the patches upstream, so what we can do is a PR with a big revert of all IPC4 code in topic/sof-dev, and restart with a clean slate.

@plbossart, #3402 is created.

ujfalusi commented:
@plbossart, Is there anything I should address in this PR, can you re-check and let me know?

plbossart commented on drivers/acpi/acpi_processor.c:
```diff
@@ -880,8 +880,6 @@ int acpi_processor_evaluate_cst(acpi_handle handle, u32 cpu,
 	}
 
-	acpi_handle_info(handle, "Found %d idle states\n", last_index);
-
```
did you intend this to be reviewed @ujfalusi ?

ujfalusi replied:
Oops

plbossart commented on sound/soc/sof/sof-priv.h:
```diff
@@ -378,11 +378,25 @@ struct ipc_pcm_ops;
+	int (*set_get_data)(struct snd_sof_dev *sdev, void *data, size_t data_bytes,
+			    bool set);
+
+	int (*get_reply)(struct snd_sof_dev *sdev);
```
are they all mandatory, or some mandatory and some optional?

Later in the series you check for get_reply as mandatory, but not the others?

ujfalusi replied:
They should be all mandatory, I'll clarify this.

plbossart commented on sound/soc/sof/ipc.c:
```diff
@@ -1061,8 +1015,8 @@ struct snd_sof_ipc *snd_sof_ipc_init(struct snd_sof_dev *sdev)
 	/* check for mandatory ops */
-	if (!ipc->ops->pcm || !ipc->ops->tplg || !ipc->ops->tplg->widget ||
-	    !ipc->ops->tplg->control) {
+	if (!ipc->ops->get_reply || !ipc->ops->pcm || !ipc->ops->tplg ||
```
see earlier comment, why is get_reply mandatory and not the others? tx_msg, rx_msg and set_get_data are missing. is this intentional, and if yes a comment would help.

ujfalusi replied:
That's a good question, they should all be mandatory.

plbossart commented on sound/soc/sof/ipc.c:
```diff
@@ -1015,8 +760,9 @@ struct snd_sof_ipc *snd_sof_ipc_init(struct snd_sof_dev *sdev)
-	if (!ipc->ops->get_reply || !ipc->ops->pcm || !ipc->ops->tplg ||
-	    !ipc->ops->tplg->widget || !ipc->ops->tplg->control) {
+	if (!ipc->ops->get_reply || !ipc->ops->tx_msg || !ipc->ops->set_get_data ||
+	    !ipc->ops->pcm || !ipc->ops->tplg || !ipc->ops->tplg->widget ||
```
ah, that answers partially to my previous comment but I still don't see rx_msg?

ujfalusi replied:
Right, now I got it.

First I have switched to get_reply ops, with that change the given ops is made mandatory.
This patch switches to the tx_msg and set_get_data ops, so from this point these are mandatory.

The rx_msg still has full implementation in ipc.c, so it is not mandatory, but if I'm not mistaken when we switch to the rx_msg ops, then it is also going to be mandatory.

plbossart commented on sound/soc/sof/sof-audio.h:
```diff
@@ -86,6 +86,7 @@ struct ipc_tplg_control_ops {
 			     const unsigned int __user *binary_data, unsigned int size);
 	/* update control data based on notification from the DSP */
 	void (*update)(struct snd_sof_dev *sdev, void *ipc_control_message);
+	int (*widget_kcontrol_setup)(struct snd_sof_dev *sdev, struct snd_sof_widget *swidget);
```
mark it as optional.

ujfalusi replied:
OK.

plbossart reviewed (changes requested):
looks mostly good @ujfalusi, I added a set of nitpicks below - mainly to clarify for all the ops callbacks which ones are mandatory and which ones are optional in the definition. That will help reviewers.

ujfalusi commented on sound/soc/sof/ipc.c:
```diff
@@ -761,8 +309,8 @@ struct snd_sof_ipc *snd_sof_ipc_init(struct snd_sof_dev *sdev)
 	if (!ipc->ops->get_reply || !ipc->ops->tx_msg || !ipc->ops->set_get_data ||
-	    !ipc->ops->pcm || !ipc->ops->tplg || !ipc->ops->tplg->widget ||
-	    !ipc->ops->tplg->control) {
+	    !ipc->ops->rx_msg || !ipc->ops->pcm || !ipc->ops->tplg ||
```
@plbossart, at this point the rx_msg is mandatory.

kv2019i commented on sound/soc/sof/sof-priv.h:
```diff
@@ -378,11 +378,25 @@ struct ipc_pcm_ops;
  * @tplg:	Pointer to IPC-specific topology ops
```
typo in commit message: s/get_reply()ops/get_reply() ops/

kv2019i commented on sound/soc/sof/ipc3.c:
```diff
@@ -316,6 +318,105 @@ static int sof_ipc3_tx_msg(struct snd_sof_dev *sdev, void *msg_data, size_t msg_
 	return ret;
 }
 
+static int sof_ipc3_set_get_data(struct snd_sof_dev *sdev, void *data, size_t data_bytes,
```
I guess this is also based on existing implementation, just restructured? Might be good to mention in the git commit message.

ujfalusi replied:
Yes, it is based on that, but split up to be able to have this kind of generic ops which takes care of the splitting and sending.

I'll update the commit message

kv2019i reviewed (commented):
A quick drive-by-review for the series and looks good for me. The git commit messages are helpful to understand the code moves, making this easier to review.

ujfalusi commented:
Changes since v4:
- commit messages updated as suggested
- comments expanded where requested
- Introduce rx_data within snd_sof_ipc_msg struct to carry the message from low level platform code to generic higher level IPC code
- re-structure the ipc ops checking
- use separate commit to add the ipc message handling ops checks to avoid confusion

ranj063 reviewed (dismissed):
looks great @ujfalusi. Thanks!

ranj063 commented:
checkpatch caught an error @ujfalusi . if you're going to fix that I have another nit-pick. you have "setup" as a single word is some comments. mIght be good to split them up as set up

ujfalusi commented:
> Changes since v4:
>     * commit messages updated as suggested
>     * comments expanded where requested
>     * Introduce rx_data within snd_sof_ipc_msg struct to carry the message from low level platform code to generic higher level IPC code

This is actually not part of this PR, it is introduced in the upcoming PR which will add back the IPC4 support.

> 
>     * re-structure the ipc ops checking
>     * use separate commit to add the ipc message handling ops checks to avoid confusion

ujfalusi commented:
> checkpatch caught an error @ujfalusi.

There is one warning with the repeated 'the', I'll fix it for sure.

> if you're going to fix that I have another nit-pick. you have "setup" as a single word is some comments. mIght be good to split them up as set up

OK, let me go through the patches if I can also spot it.

ujfalusi commented:
Changes since v5:
- Fix the checkpatch warning (repeated 'the')
- correct the spelling of set up from 'setup' to 'set up'

plbossart commented on sound/soc/sof/ipc.c:
```diff
@@ -340,6 +340,56 @@ static int sof_ipc_tx_message_unlocked(struct snd_sof_ipc *ipc,
+ * @sdev:		pointer to SOF core device struct
+ * @msg_data:		pointer to a message to send
+ * @msg_bytes:		number of bytes in the message
+ * @reply_bytes:	number of bytes available for the reply
```
nit-pick: available where? at the end of msg_data, after msg_bytes?

ujfalusi replied:
@plbossart, neither ;)
This is a helper function to set up the `struct snd_sof_ipc_msg` at `sdev->ipc->msg` with the needed parameters for sending a message and then also make a call to the platform code make the transfer.
`msg_data` and `msg_bytes` is the message we send and it is passed via the `msg`, for the reply we only tell the platform code the maximum available space given by the caller to receive back the reply, the pointer to the reply is not given as the reply content handover is in the IPC specific code (`sof_ipc3_get_reply`, AMD have this open-coded).

The comment is actually correct, it is the bytes available for the reply.

plbossart replied:
not following the explanation, sorry. the caller provides the number of bytes for the reply, but not the pointer for the reply?

ujfalusi replied:
> not following the explanation, sorry. the caller provides the number of bytes for the reply, but not the pointer for the reply?

The `sof_ipc_send_msg()` is a helper function to fill the `sdev->ipc->msg` struct as it is a generic struct. We could have this as a big #define or strict inline function if we want, or just duplicate it for every IPC version.

The caller is the one who calls the `sof_ipc_tx_message()`, there we have the reply pointer and the size allocated for it. This caller provided reply data pointer is only used when we got the reply into the sdev->ipc->msg.reply_data (which is allocated at probe time) and it is processed in the IPC code and finally we copy from the intermediate storage to the user provided buffer.

The message sending itself does not need the caller provided reply buffer, the sending itself does not even needs to know the size the caller provided.
The size is needed when we got the interrupt that the message has been processed, and we can read out the reply to the temporary buffer. Here we check the caller's provided size for the reply and do minimal copy and size checks.
As a last step the we copy the data to the caller's buffer, but it is done via the completion handling in the tx path (we wait for the signal that we got the reply).

It is complicated and simple at the same time, but I really don't know how else it could be handled without major changes in the logic.

plbossart replied:
thanks for the explanations @ujfalusi. I never realized that we had an intermediate msg->reply_data intermediate buffer and a memcpy.

plbossart commented on sound/soc/sof/sof-priv.h:
```diff
@@ -378,11 +378,32 @@ struct ipc_pcm_ops;
+ *		function may split up the 'large' message and use the @tx_msg
+ *		path to transfer individual chunks, or use other means to transfer
+ *		the message.
+ * @get_reply:	Function pointer for fetching the reply to
```
reply to what ? missing word here?

ujfalusi replied:
No, the `rx_msg` got in the middle.

plbossart replied:
I don't get the answer but that can be ignored or fixed with a fixup.

plbossart commented on sound/soc/sof/ipc.c:
```diff
@@ -1039,6 +1039,7 @@ int sof_ipc_init_msg_memory(struct snd_sof_dev *sdev)
 struct snd_sof_ipc *snd_sof_ipc_init(struct snd_sof_dev *sdev)
 {
 	struct snd_sof_ipc *ipc;
+	const struct ipc_ops *ops;
```
I just realized we don't have an 'sof' prefix for the structure. Should we use sof_ipc_ops?

ujfalusi replied:
Yes, I think it is warranted

ujfalusi replied:
Hrm, then should we rename all the other ops struct, like
struct ipc_pcm_ops
struct ipc_tplg_ops
struct ipc_tplg_control_ops

@ranj063, @plbossart ?

plbossart replied:
yes, but I wonder if we can do this easily with fixups, or it's too late already and we just accept this and move on.

ujfalusi replied:
I can add the fixups to this series as I did with the sof_ipc_ops, if that's what we want. Separate PR will conflict with everything ;)

plbossart replied:
ok, let's do a fixups series then.

ujfalusi replied:
@plbossart, I have taken a look on how to do a fixup for the missing `sof_` prefixes in the existing ops, but it is not going to be nice. We can not fix up by individual commits as we will break compilation, There were also a bit of churn in the files, so the fixup is not going to apply cleanly. Let me see.

I'll drop the fixup from this PR and create a new one. They will conflict big time, so one will need to be rebased depending on the order.

ujfalusi commented:
Changes since v6:
- fixup patch to rename `struct ipc_ops` to `struct sof_ipc_ops`
- Comments updated as requested by @plbossart

plbossart commented on sound/soc/sof/ipc3-ops.h:
```diff
@@ -16,6 +16,6 @@
 extern const struct ipc_tplg_ops ipc3_tplg_ops;
 extern const struct ipc_tplg_control_ops tplg_ipc3_control_ops;
-extern const struct ipc_ops ipc3_ops;
+extern const struct sof_ipc_ops ipc3_ops;
```
we probably need a fixup for all of ipc_pcm_ops, ipc_tplg_ops, ipc_tplg_controls_ops?

plbossart reviewed (approved):
nice job @ujfalusi thanks for bearing will all my questions :-)

ujfalusi commented:
Changes since v7:
- Drop the fixup patch for sof_ipc_ops, it is unrelated to this series
